### PR TITLE
tables: when importing data, table.AddRecord has no needs to use temp BufferStore.

### PR DESCRIPTION
**Pull request by winkyao**
The benchmark result is mutable, and is influenced by some subtle reasons, so I run several times to compare this pr and master:

```
this pr:
[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 54.250 sec

[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 49.616 sec

[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 50.948 sec

[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 48.977 sec

master:
[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 58.613 sec

[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 65.942 sec

[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 53.582 sec

[direct] kv num = 24576000 / kv size = 2724864000
[direct] cost avg = 65.376 sec
```

In general, this pr has obvious improvement, in the flamegraph, InsertExec save about 5% time.

### CODE REVIEW

**Comment by winkyao** [comment]
@shenli @coocood @zz-jason PTAL

File: table/tables/tables.go [hunk 1]
```diff
@@ -326,6 +326,22 @@ func adjustRowValuesBuf(sessVars *variable.SessionVars, rowLen int) {
...
+// Just add the kv to transaction's membuf directly. the returned *kv.BufferStore avoids
+// type assert in bs.SaveTo(txn).
+func (t *Table) getRollbackableMemStore(ctx context.Context) (kv.RetrieverMutator, *kv.BufferStore) {
+	if ctx.GetSessionVars().ImportingData {
+		return ctx.Txn(), nil
+	}
+
+	bs := ctx.GetSessionVars().BufStore
+	if bs == nil {
+		bs = kv.NewBufferStore(ctx.Txn(), kv.DefaultTxnMembufCap)
```
Comment (shenli) [inline table/tables/tables.go]: ctx.GetSessionVars().BufStore = bs ?

File: table/tables/tables.go [hunk 2]
```diff
@@ -326,6 +326,22 @@ func adjustRowValuesBuf(sessVars *variable.SessionVars, rowLen int) {
...
+	}
+	bs.Reset()
```
Comment (shenli) [inline table/tables/tables.go]: Should we put the Reset() in a `else` clause?

File: table/tables/tables.go [hunk 3]
```diff
@@ -394,8 +406,10 @@ func (t *Table) AddRecord(ctx context.Context, r []types.Datum, skipHandleCheck
 	if err = txn.Set(key, value); err != nil {
 		return 0, errors.Trace(err)
 	}
-	if err = bs.SaveTo(txn); err != nil {
-		return 0, errors.Trace(err)
+	if !sessVars.ImportingData && bs != nil {
```
Comment (shenli) [inline table/tables/tables.go]: If !sessVars.ImportingData, the bs should never be nil.

File: table/tables/tables.go (hunk 2 above)
Comment (winkyao) [reply table/tables/tables.go]: no, the session.Buf also needs to reset

File: table/tables/tables.go (hunk 1 above)
Comment (winkyao) [reply table/tables/tables.go]: ctx.GetSessionVars().BufStore is new in InsertExec.exec, if it is nil, means we are in other executor and not use sessVars.BufStore yet. I will refactor these buffers code to make it clear and graceful.

File: table/tables/tables.go [hunk 4]
```diff
@@ -326,6 +326,23 @@ func adjustRowValuesBuf(sessVars *variable.SessionVars, rowLen int) {
 	sessVars.AddRowValues = sessVars.AddRowValues[:adjustLen]
 }
 
+// getRollbackableMemStore get a rollbackable BufferStore, when we are importing data,
+// Just add the kv to transaction's membuf directly. the returned *kv.BufferStore avoids
+// type assert in bs.SaveTo(txn).
+func (t *Table) getRollbackableMemStore(ctx context.Context) (kv.RetrieverMutator, *kv.BufferStore) {
```
Comment (coocood) [inline table/tables/tables.go]: I think a type assertion is fine.
Return only `RetrieverMutator` is easier to read.

**Comment by zz-jason** [comment]
LGTM

**Pull request merged**